File: ts-pr/55/typescript/code-quality-logging/user_service.ts
class UserService {
    private users: string[] = [];

    addUser(username: string, password: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`Adding user: ${username} with password: ${password}`);
        this.users.push(username);
        // {/fact}
    }

    safeAddUser(username: string, password: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        console.log(`Adding user: ${username}`);
        this.users.push(username);
        // {/fact}
    }

    authenticateUser(username: string, password: string): boolean {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`Authentication attempt for ${username} with password ${password}`);
        return true;
        // {/fact}
    }

    safeAuthenticateUser(username: string, password: string): boolean {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        console.log(`Authentication attempt for user: ${username}`);
        return true;
        // {/fact}
    }

    updatePassword(username: string, oldPassword: string, newPassword: string): boolean {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`Password update for ${username}: ${oldPassword} -> ${newPassword}`);
        return true;
        // {/fact}
    }

    safeUpdatePassword(username: string, oldPassword: string, newPassword: string): boolean {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        console.log(`Password updated for user: ${username}`);
        return true;
        // {/fact}
    }
}

const service = new UserService();
service.addUser('john_doe', 'secret123');
service.authenticateUser('john_doe', 'secret123');

File: go-pr/48/go/code-quality-logging/user_service.go
package main

import (
	"fmt"
	"log"
)

type UserService struct {
	users []string
}

func (us *UserService) AddUser(username, password string) {
	// {fact rule=code-quality-logging@v1.0 defects=1}
	fmt.Printf("Adding user: %s with password: %s\n", username, password)
	us.users = append(us.users, username)
	// {/fact}
}

func (us *UserService) SafeAddUser(username, password string) {
	// {fact rule=code-quality-logging@v1.0 defects=0}
	log.Printf("Adding user: %s\n", username)
	us.users = append(us.users, username)
	// {/fact}
}

func (us *UserService) AuthenticateUser(username, password string) bool {
	// {fact rule=code-quality-logging@v1.0 defects=1}
	fmt.Printf("Authentication attempt for %s with password %s\n", username, password)
	return true
	// {/fact}
}

func (us *UserService) SafeAuthenticateUser(username, password string) bool {
	// {fact rule=code-quality-logging@v1.0 defects=0}
	log.Printf("Authentication attempt for user: %s\n", username)
	return true
	// {/fact}
}

func main() {
	service := &UserService{}
	service.AddUser("john_doe", "secret123")
	service.AuthenticateUser("john_doe", "secret123")
}